### PR TITLE
Plat 602 keycloak integration

### PR DESCRIPTION
**Pull request by nour-borgi**
In this PR: 

- Add keycloak startegy (using passport-openidconnect)

### CODE REVIEW

File: src/protocols/openid.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    identifier: identifier
+  })
+    .then(function (passport) {
+      if (!req.user) {
+        // Scenario: A (new) user is attempting to sign up using a third-party
+        //           authentication provider.
+        // Action:   Create/update the user and assign them a passport.
+        if (!passport) {
+          return createOrUpdateUser(user)
+            .then(function (_user) {
```
Comment (bradsawadye) [inline src/protocols/openid.js:101]: Minor - rename the parameter for the function to something else. The underscore is used to indicate that a parameter is not used in the function.

File: src/protocols/openid.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  user.token = null
+  user.tokenType = null
+  user.locked = false
+  user.expiry = null
+
+  return {user, identifier: sub}
+}
+
+function createOrUpdateUser(user) {
+  return UserModelAPI.findOne({email: user.email}).then(function (_user) {
```
Comment (bradsawadye) [inline src/protocols/openid.js:55]: Minor - rename the parameter for the function to something else. The underscore is used to indicate that a parameter is not used in the function.

File: src/model/users.js [hunk 3]
```diff
@@ -153,18 +152,19 @@ export const updateUser = async function (_user) {
  *
  */
 export const updateTokenUser = async function (_user) {
```
[the remark this answers is not in the document]
Comment (nour-borgi) [reply src/model/users.js]: cool done :)

File: src/protocols/openid.js (hunk 2 above)
Comment (nour-borgi) [reply src/protocols/openid.js]: cool done :)

File: src/protocols/openid.js (hunk 1 above)
Comment (nour-borgi) [reply src/protocols/openid.js]: cool done :)

File: config/default.json [hunk 4]
```diff
@@ -42,7 +42,14 @@
...
     "truncateAppend": "\n[truncated ...]",
-    "authenticationTypes": ["basic", "local", "token"]
+    "authenticationTypes": ["basic", "local", "token", "openid"],
+    "keycloak": {
+      "url": "http://localhost:9088/realms/platform-realm",
+      "callbackUrl": "http://localhost:9000",
+      "clientId": "openhim-oauth",
+      "clientSecret": "tZKfEbWf0Ka5HBNZwFrdSyQH2xT1sNMR",
+      "scope": "openid email profile offline_access roles"
+    }
```
Comment (rcrichton) [inline config/default.json:52]: Please add descriptions of these to the config.md file.

Comment (nour-borgi) [reply config/default.json:52]: sure. Could you review the base branch again, addressing the feedback put the review as stale?

Comment (nour-borgi) [reply config/default.json:52]: I also merged the test fix branch into that branch

File: config/default.json [hunk 5]
```diff
@@ -42,7 +42,14 @@
     "maxPayloadSizeMB": 50,
     "truncateSize": 15000,
     "truncateAppend": "\n[truncated ...]",
-    "authenticationTypes": ["basic", "local", "token"]
+    "authenticationTypes": ["basic", "local", "token", "openid"],
+    "keycloak": {
```
Comment (rcrichton) [inline config/default.json:46]: The openid auth type isn't specifically restricted to keycloak. I wonder if we should rename this section to `oauth` or something similar. Then we should also not make so much mention of keycloak in the rest of the code. I don't see anything that ties us to keycloak right now.

**Review by rcrichton — CHANGES_REQUESTED** [review]

File: config/default.json (hunk 5 above)
Comment (nour-borgi) [reply config/default.json]: cool done :)

File: config/config.md [hunk 6]
```diff
@@ -71,8 +71,22 @@ The following config option are provided by the OpenHIM. All of these options ha
     // * "local" means through the UI with hitting "/authentication/local" endpoint with username and password, 
     // this will create a session for the user and set cookies in the browser.
     // * "basic" means with basic auth either through browser or postman by giving also username and password.
+    // * "openid" means with a third party authentication provider (e.g. keycloak).
     // * [Deprecated] "token" means that a request should provide in the header an 'auth-token', 'auth-salt' and 'auth-ts' to be authenticated.
-    "authenicationTypes": ["token"]
+    "authenicationTypes": ["token"],
+    // Keycloak configuration needed for the authentication
+    "keycloak": {
```
Comment (rcrichton) [inline config/config.md:78]: This one still needs to change.

**Review by rcrichton — APPROVED** [review]
Looks good to me, just one small comment.

File: config/config.md (hunk 6 above)
Comment (nour-borgi) [reply config/config.md]: Thanks, done :)

**Review by rcrichton — APPROVED** [review]
LGTM!

**Pull request merged**